Update InPlacePodVerticalScaling docs for v1.33 beta (#50290)
kubernetes/website: content/en/docs/tasks/configure-pod-container/resize-container-resources.md

**Key Concepts:**

* **Desired Resources:** A container's `spec.containers[*].resources` represent the *desired* resources for the container, and are mutable for CPU and memory.
* **Actual Resources:** The `status.containerStatuses[*].resources` field reflects the resources *currently applied* for a running container. For containers that haven't started or were restarted, it reflects the resources allocated upon their next start.
* **Triggering a Resize:** You request a resize by updating the desired `requests` and `limits` in the Pod's specification. This is typically done using `kubectl patch`, `kubectl apply`, or `kubectl edit` targeting the Pod's `resize` subresource. When the desired resources don't match the allocated resources, the Kubelet will attempt to resize the container.
* **Allocated Resources (Advanced):** The `status.containerStatuses[*].allocatedResources` field tracks resource values confirmed by the Kubelet, primarily used for internal scheduling logic. For most monitoring and validation purposes, focus on `status.containerStatuses[*].resources`.

> If a node has pods with a pending or incomplete resize (see [Pod Resize Status](#pod-resize-status) below), the scheduler uses the *maximum* of a container's desired requests, allocated requests, and actual requests from the status when making scheduling decisions.

Reviewer: ```suggestion
If a node has pods with a pending or incomplete resize (see [Pod Resize Status](#pod-resize-status) below), the {{< glossary_tooltip text="scheduler" term_id="kube-scheduler" >}}
uses the *maximum* of a container's desired requests, allocated requests, and actual requests from the status when making scheduling decisions.
```

* **Triggering a Resize:** You request a resize by updating the desired `requests` and `limits` in the Pod's specification. This is typically done using `kubectl patch`, `kubectl apply`, or `kubectl edit` targeting the Pod's `resize` subresource. When the desired resources don't match the allocated resources, the Kubelet will attempt to resize the container.
* **Allocated Resources (Advanced):** The `status.containerStatuses[*].allocatedResources` field tracks resource values confirmed by the Kubelet, primarily used for internal scheduling logic. For most monitoring and validation purposes, focus on `status.containerStatuses[*].resources`.

If a node has pods with a pending or incomplete resize (see [Pod Resize Status](#pod-resize-status) below), the scheduler uses the *maximum* of a container's desired requests, allocated requests, and actual requests from the status when making scheduling decisions.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

The `InPlacePodVerticalScaling` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) must be enabled

> for your control plane and for all nodes in your cluster.

Reviewer: This is still true. I wouldn't drop it yet.

Author: Ack. I wasn't sure if the feature gate header was sufficient to cover this.

> ## Pod Resize Status

Reviewer: Per style guide, we recommend sentence case. For things enabled by default, we like to have docs at almost GA quality.
```suggestion
## Pod resize status
```

> ## Container Resize Policies

Reviewer: Per style guide, we recommend sentence case. For things enabled by default, we like to have docs at almost GA quality.
```suggestion
## Container resize policies
```

> You can control whether a container should be restarted when resizing by setting `resizePolicy` within the container specification. This allows fine-grained control based on resource type (CPU or memory).

Reviewer: What about extended resources (eg from [device plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/))? The docs should make the story clear for those.

Author: This is covered under key-concepts (only CPU & Memory are mutable) and limitations. Do you have any suggestions of how to raise the visibility of this?

> In-place resource resizing currently has the following limitations:

Reviewer: ```suggestion
For Kubernetes {{< skew currentVersion >}}, in-place resource resizing has the following limitations:
```

* **Resource Types:** Only CPU and memory resources can be resized.
* **Memory Decrease:** Memory limits _cannot be decreased_ unless the `resizePolicy` for memory is `RestartContainer`. Memory requests can generally be decreased.

> * **QoS Class:** The Pod's original [Quality of Service (QoS) class](/docs/tasks/configure-pod-container/quality-service-pod/) (Guaranteed, Burstable, BestEffort) is determined at creation and cannot be changed by a resize. The resized resource values must still adhere to the rules of the original QoS class:

Reviewer: ```suggestion
* **QoS Class:** The Pod's original [Quality of Service (QoS) class](/docs/concepts/workloads/pods/pod-qos/) (Guaranteed, Burstable, or BestEffort) is determined at creation and **cannot** be changed by a resize.
  The resized resource values must still adhere to the rules of the original QoS class:
```

    * *Guaranteed*: Requests must continue to equal limits for both CPU and memory after resizing.
    * *Burstable*: Requests and limits cannot become equal for *both* CPU and memory simultaneously (as this would change it to Guaranteed).
    * *BestEffort*: Resource requirements (`requests` or `limits`) cannot be added (as this would change it to Burstable or Guaranteed).

> * **Container Types:** Non-restartable Init Containers and Ephemeral Containers cannot be resized. [Sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) can be resized.

Reviewer: ```suggestion
* **Container Types:** Non-restartable {{< glossary_tooltip text="init containers" term_id="init-container" >}} and {{< glossary_tooltip text="ephemeral containers" term_id="ephemeral-container" >}} cannot be resized.
  {{< glossary_tooltip text="Sidecar containers" term_id="sidecar-container" >}}(/docs/concepts/workloads/pods/sidecar-containers/) can be resized.
```

* **Resource Removal:** Resource requests and limits cannot be entirely removed once set; they can only be changed to different values.

> * **Operating System:** Windows pods do not support in-place resize.

Reviewer: Is this true:
```suggestion
* **Operating System:** Pods with the OS set to Windows do not support in-place resize. If you attempt to resize a Pod that is bound
  to a Windows node, you also cannot resize it. However, you can attempt to resize a Pod that uses a Windows container, provided it has been created but is not yet bound to a node.
```
?

Author: We attempt to enforce this in admission, so the apiserver will reject the request if the OS is set to Windows. In the second case, the request will go through, but the resize is marked as infeasible, so it's still not supported. I didn't distinguish in this list which limitations are enforced in admission and which are enforced by the Kubelet. That level of detail seems unnecessary?

Author: Is this better?
```suggestion
* **Operating System:** Pods on Windows nodes cannot be resized.
```

Author: Oh, I missed this part

> However, you can attempt to resize a Pod that uses a Windows container, provided it has been created but is not yet bound to a node.

That is technically true, but IMO we shouldn't document it. It's an edge case that we don't want to encourage.

* **Node Policies:** Pods managed by [static CPU or Memory manager policies](/docs/tasks/administer-cluster/cpu-management-policies/) cannot be resized in-place.

> * **Swap:** Pods utilizing [swap memory](/docs/concepts/architecture/nodes/#swap-memory) cannot resize memory requests unless the `resizePolicy` for memory is `RestartContainer`.

Reviewer: Is this true?
```suggestion
* **Swap:** Pods utilizing [swap memory](/docs/concepts/architecture/nodes/#swap-memory) can only resize memory requests if the `resizePolicy` for memory is `RestartContainer`. If the `resizePolicy` is set to something else, the API server accepts the resize request but it will not succeed.
```

Author: Again, do you think it's necessary to describe the enforcement mechanism here? Here's the full list:

Enforced in API admission:
- Resource type
- QoS class
- Container types
- Windows pods (only if OS name is windows)
- Resource removal

Enforced at allocation time (resize marked Infeasible):
- Static CPU / memory
- Swap
- node Capacity (not listed here)
- Windows (anything that slipped through admission)

These restrictions might be relaxed in future Kubernetes versions.

## Example 1: Resizing CPU without Restart

> First, create a Pod designed for in-place CPU resize and restart-required memory resize.

Reviewer: What namespace is this Pod in? Do you need to create a namespace first?

Author: I was thinking they could just use the default namespace, since it's just a single resource being created. Is it standard practice to create a dedicated namespace for all tutorials?

{{% code_sample file="pods/resource/pod-resize.yaml" %}}

Create the pod:

```shell
kubectl create -f pod-resize.yaml
```

This pod starts in the Guaranteed QoS class. Verify its initial state:

```shell
# Wait a moment for the pod to be running
kubectl get pod resize-demo --output=yaml
```

Observe the `spec.containers[0].resources` and `status.containerStatuses[0].resources`. They should match the manifest (700m CPU, 200Mi memory). Note the `status.containerStatuses[0].restartCount` (should be 0).

> Now, let's increase the CPU request and limit to `800m`. We use `kubectl patch` with the `--subresource resize` flag.

Reviewer: We [avoid using "we"](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we)
```suggestion
Now, increase the CPU request and limit to `800m`. You use `kubectl patch` with the `--subresource resize` flag.
```

```shell

> kubectl patch pod resize-demo --subresource resize --patch \

Reviewer: What namespace is change happening in? Do you need to create a namespace first?

Author: (see above)

  '{"spec":{"containers":[{"name":"pause", "resources":{"requests":{"cpu":"800m"}, "limits":{"cpu":"800m"}}}]}}'

# Alternative methods:
# kubectl -n qos-example edit pod resize-demo --subresource resize
# kubectl -n qos-example apply -f <updated-manifest> --subresource resize
```

{{< note >}}

> The `--subresource resize` flag requires `kubectl` client version v1.32.0 or later. Older versions will report an `invalid subresource` error.

Reviewer: ```suggestion
The `--subresource resize` command line aregument requires `kubectl` client version v1.32.0 or later. Older versions will report an `invalid subresource` error.
```

```

You should see:
* `spec.containers[0].resources` now shows `cpu: 800m`.
* `status.containerStatuses[0].resources` also shows `cpu: 800m`, indicating the resize was successful on the node.
* `status.containerStatuses[0].restartCount` remains `0`, because the CPU `resizePolicy` was `NotRequired`.

## Example 2: Resizing Memory with Restart

Now, let's resize the memory for the *same* pod. We'll increase it to `300Mi`. Since the memory `resizePolicy` is `RestartContainer`, we expect the container to restart.

```shell
kubectl patch pod resize-demo --subresource resize --patch \
  '{"spec":{"containers":[{"name":"pause", "resources":{"requests":{"memory":"300Mi"}, "limits":{"memory":"300Mi"}}}]}}'
```

Check the pod status shortly after patching:

```shell
kubectl get pod resize-demo --output=yaml
```

You should now observe:
* `spec.containers[0].resources` shows `memory: 300Mi`.
* `status.containerStatuses[0].resources` also shows `memory: 300Mi`.
* `status.containerStatuses[0].restartCount` has increased to `1` (or more, if restarts happened previously), indicating the container was restarted to apply the memory change.

## Troubleshooting: Infeasible Resize Request

> Let's try requesting an unreasonable amount of CPU, like 1000 full cores (written as `"1000"` instead of `"1000m"` for millicores), which likely exceeds node capacity.

Reviewer: ```suggestion
Next, try requesting an unreasonable amount of CPU, such as 1000 full cores (written as `"1000"` instead of `"1000m"` for millicores), which likely exceeds node capacity.
```

    started: true
```shell
kubectl get pod resize-demo --output=yaml
```

Observe that the `resources` in the `containerStatuses` have been updated to reflect the new desired
CPU requests. This indicates that node was able to accommodate the increased CPU resource needs,
and the new CPU resources have been applied. The Container's `restartCount` remains unchanged,
indicating that container's CPU resources were resized without restarting the container.
You'll see changes indicating the problem:

* The `spec.containers[0].resources` reflects the *desired* state (`cpu: "1000"`).
* A condition with `type: PodResizePending` and `reason: Infeasible` was added to the Pod.
* The condition's `message` will explain why (`Node didn't have enough capacity: cpu, requested: 800000, capacity: ...`)
* Crucially, `status.containerStatuses[0].resources` will *still show the previous values* (`cpu: 800m`, `memory: 300Mi`), because the infeasible resize was not applied by the Kubelet.
* The `restartCount` will not have changed due to this failed attempt.

To fix this, you would need to patch the pod again with feasible resource values.

## Clean up

Delete your namespace:
Delete the pod:

```shell

> kubectl delete namespace qos-example

Reviewer: Does the namespace need removing or not?

title: Resize CPU and Memory Resources assigned to Containers
content_type: task
weight: 30
min-kubernetes-server-version: 1.33
---

<!-- overview -->

{{< feature-state feature_gate_name="InPlacePodVerticalScaling" >}}

This page explains how to change the CPU and memory resource requests and limits assigned to a container *without recreating the Pod*.

> Traditionally, changing a Pod's resource requirements necessitated deleting the existing Pod and creating a replacement, often managed by a [workload controller](/docs/concepts/workloads/controllers/). In-place resizing allows containers within a running Pod to be allocated more or less CPU/memory, potentially avoiding application disruption.

Reviewer 2: Suggest: `In-place resizing allows containers within a running Pod to be allocated more or less CPU/memory, potentially avoiding application disruption.` -> `In-place Pod Resize allows changing the CPU/memory allocation of container(s) within a running Pod while potentially avoiding application disruption.`

**Key Concepts:**

* **Desired Resources:** A container's `spec.containers[*].resources` represent the *desired* resources for the container, and are mutable for CPU and memory.

> * **Actual Resources:** The `status.containerStatuses[*].resources` field reflects the resources *currently applied* for a running container. For containers that haven't started or were restarted, it reflects the resources allocated upon their next start.

Reviewer 2: ` currently applied for a running container` -> `currently configured on a running container`

> * **Triggering a Resize:** You request a resize by updating the desired `requests` and `limits` in the Pod's specification. This is typically done using `kubectl patch`, `kubectl apply`, or `kubectl edit` targeting the Pod's `resize` subresource. When the desired resources don't match the allocated resources, the Kubelet will attempt to resize the container.

Reviewer 2: `You request a resize...` -> `You can request a resize...`

    - resourceName: cpu
      restartPolicy: NotRequired
    - resourceName: memory
      restartPolicy: RestartContainer
```

* `NotRequired`: (Default) Apply the resource change to the running container without restarting it.
* `RestartContainer`: Restart the container to apply the new resource values. This is often necessary for memory changes because many applications and runtimes cannot adjust their memory allocation dynamically.

If `resizePolicy[*].restartPolicy` is not specified for a resource, it defaults to `NotRequired`.

{{< note >}}
If a Pod's overall `restartPolicy` is `Never`, then any container `resizePolicy` must be `NotRequired` for all resources. You cannot configure a resize policy that would require a restart in such Pods.
{{< /note >}}

**Example Scenario:**

Consider a container configured with `restartPolicy: NotRequired` for CPU and `restartPolicy: RestartContainer` for memory.
* If only CPU resources are changed, the container is resized in-place.
* If only memory resources are changed, the container is restarted.
* If *both* CPU and memory resources are changed simultaneously, the container is restarted (due to the memory policy).

## Limitations

> For Kubernetes {{< skew currentVersion >}}, in-place resource resizing has the following limitations:

Reviewer 2: suggest changing `in-place resource resizing` -> `in-place pod resize` in order to be consistent on what we want to call this feature.

Author: Flipped it around to avoid naming the feature instead.

Reviewer 2: Looks good! thanks
